README.md
# UBJSON encoder/decoder

[![npm](https://img.shields.io/npm/v/@shelacek/ubjson.svg)](https://www.npmjs.com/package/@shelacek/ubjson)
[![npm bundle size (minified + gzip)](https://img.shields.io/bundlephobia/minzip/@shelacek/ubjson.svg)](https://bundlephobia.com/result?p=@shelacek/ubjson)
[![Bitbucket Pipelines](https://img.shields.io/bitbucket/pipelines/shelacek/ubjson.svg)](https://bitbucket.org/shelacek/ubjson/addon/pipelines/home)
[![Codecov](https://img.shields.io/codecov/c/bb/shelacek/ubjson.svg)](https://codecov.io/bb/shelacek/ubjson)

*Lightweight, quick and dirty UBJSON encoder/decoder for the browser and node.js.*

This library implements [Universal Binary JSON] Draft 12 encoder/decoder in JavaScript.

> ⚠ There are some changes between unstable versions and version 1.0.0. Please have a look
> at the [CHANGELOG.md].

Encoding and decoding working as far as I know. If you encounter an bug, please [create an issue].

[Universal Binary JSON]: http://ubjson.org/
[create an issue]: https://bitbucket.org/shelacek/ubjson/issues
[CHANGELOG.md]: https://bitbucket.org/shelacek/ubjson/src/master/CHANGELOG.md


## Example (browser)

```js
import { Ubjson } from '@shelacek/ubjson';
const buffer = Ubjson.encode({ hello: 'world', from: ['UBJSON'] });
const obj = Ubjson.decode(buffer);
console.log(obj); // { hello: 'world', from: ['UBJSON'] }
```


## Example (node.js)

```js
const ubjson = require('@shelacek/ubjson');
const buffer = ubjson.encode({ hello: 'world', from: ['UBJSON'] });
const obj = ubjson.decode(buffer);
console.log(obj); // { hello: 'world', from: ['UBJSON'] }
```


## API

> Note: For full API and exported symbols, please see typings at
> https://bitbucket.org/shelacek/ubjson/src/master/src/types.d.ts.


### `Ubjson.encode(value, [options])`

- `value: any` - input value/object/array to serialize.
- `options: Object` (optional) - encoding options.
    - `options.optimizeArrays: boolean | 'onlyTypedArrays'` (default `false`) - enable use
      [optimized format] for arrays. If `'onlyTypedArrays'` is used, only *TypedArrays* use strongly
      typed container.
    - `options.optimizeObjects: boolean` (default `false`) - enable use [optimized format]
      for objects.

Method returns `ArrayBuffer` with *UBJSON* data.

[optimized format]: http://ubjson.org/type-reference/container-types/#optimized-format


### `Ubjson.decode(buffer, [options])`

- `buffer: ArrayBuffer` - input buffer with *UBJSON* data.
- `options: Object` (optional) - decoding options.
    - `options.int64Handling: 'error' | 'skip' | 'raw'` (default `error`) - Handling of unsupported
      *int64* values. 'error' throws exception, 'skip' ignore that value (or key/value pair) and 'raw'
      returns Uint8Array with *int64* bytes.
    - `options.highPrecisionNumberHandling: 'error' | 'skip' | 'raw'` (default `error`) - Handling
      of unsupported *high-precision numbers*. 'error' throws exception, 'skip' ignore that value
      (or key/value pair) and 'raw' returns string represents of that number.
    - `options.useTypedArrays: boolean` (default `false`) - enable use of *TypedArrays* for strongly
      typed containers.

Method returns decoded *UBJSON* value/object/array (`any`).


## Limitations

Javascript not support [64-bit integers]&nbsp;(yet) and [high-precision numbers] as well
as the library. You can use `'raw'` option in `int64Handling`/`highPrecisionNumberHandling`
to retrive original data.

[no-op value]: http://ubjson.org/type-reference/value-types/#noop
[64-bit integers]: http://ubjson.org/type-reference/value-types/#numeric-64bit
[high-precision numbers]: http://ubjson.org/type-reference/value-types/#numeric-gt-64bit


## Compatibility

The library needs a `TextEncoder` and `TextDecoder` (or `util.TextEncoder`/`util.TextDecoder`
on node.js), support for `TypedArrays` and `class` keyword. Library should work on
*Firefox*&nbsp;>=45, *Chrome*&nbsp;>=49, *Safari*&nbsp;>=10.1, *Opera*&nbsp;>=36,
*Node*&nbsp;>=8.11.3. *Edge* needs TextEncoder/TextDecoder polyfill.

In case of interest, I can reduce requirements.


## Alternatives

There are some great alternatives:

- [Sannis/node-ubjson] - node.js (only) library, that implements draft 8.
- [artcompiler/L16] - written in asm.js and implements draft 8.
- [zentner-kyle/ubjson.js] - no documentation, but it is cjs module and implements draft 12?!

> *\[Please correct me if this information is obsolete or wrong.\]*

This library was created because I needed to work with UBJSON in browser - and there are no
reasonable draft 12 implementation :-(.

[Sannis/node-ubjson]: https://github.com/Sannis/node-ubjson
[artcompiler/L16]: https://github.com/artcompiler/L16
[zentner-kyle/ubjson.js]: https://github.com/zentner-kyle/ubjson.js


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


## [Unreleased]

### Added

- Flow typings.


## [1.0.0] - 2018-07-26

### Changed

- **BREAKING**: Exporting only named exports.

    This change enable exporting `UbjsonEncoder` and `UbjsonDecoder`, so better utilization
    of tree-shaking is now possible ([#3](https://bitbucket.org/shelacek/ubjson/issues/3)).
      
    Please change

        import Ubjson from '@shelacek/ubjson';

    to

        import { Ubjson } from '@shelacek/ubjson';
  
    Now, you can also import `UbjsonEncoder`, `UbjsonDecoder`, `encode` and `decode` ;-).

- **BREAKING**: Some options refactoring ([#2](https://bitbucket.org/shelacek/ubjson/issues/2)).
    - `UbjsonDecoderOptions.useTypedArrays` defaults to `false`. Please allow it explicitly, if you
      used this behavior previously.
    - Renamed value `onlyTypedArray` of `UbjsonEncoderOptions.optimizeArrays` to `onlyTypedArrays`.


### Fixed

- Fix encoding of empty optimized containers.
- Respect `onlyTypedArrays` on encoding of non-8bit typed arrays. Previously it was serialized
  as standard not-optimized containers.


## [0.2.3] - 2018-07-22

### Fixed

- Fix encoding of non-8bit typed arrays.


## [0.2.2] - 2018-07-21

### Fixed

- Fix handling of no-op values.
- All option's properties are now optional in typings.
- Fix encoding of big integers > `int32` ([#1](https://bitbucket.org/shelacek/ubjson/issues/1)).


## [0.2.1] - 2018-07-18

### Changed

-  `README.md` updated only.


## [0.2.0] - 2018-07-17

### Changed

- **BREAKING**: Export static `Ubjson` class as default.

    Please change

        import { Ubjson } from '@shelacek/ubjson';
        // - or -
        const ubjson = require('@shelacek/ubjson').Ubjson;

    to

        import Ubjson from '@shelacek/ubjson';
        // - or -
        const ubjson = require('@shelacek/ubjson');


### Fixed

- Take `TextEncoder`/`TextDecoder` from `require('util')`, if they are not available globally.
  This makes it possible to use the library on the *node.js* >= 8 without polyfills.


## [0.1.1] - 2018-07-17

### Fixed

- Fix inappropriate promotion of `uint8` type to `int16` for STC.
- Fix encoding of typed arrays.
- Fix encoding of strongly typed containers.


## [0.1.0] - 2018-07-17

### Added

- Initial version.


[Unreleased]: https://bitbucket.org/shelacek/ubjson/branches/compare/master..v1.0.0
[1.0.0]: https://bitbucket.org/shelacek/ubjson/branches/compare/v1.0.0..v0.2.3
[0.2.3]: https://bitbucket.org/shelacek/ubjson/branches/compare/v0.2.3..v0.2.2
[0.2.2]: https://bitbucket.org/shelacek/ubjson/branches/compare/v0.2.2..v0.2.1
[0.2.1]: https://bitbucket.org/shelacek/ubjson/branches/compare/v0.2.1..v0.2.0
[0.2.0]: https://bitbucket.org/shelacek/ubjson/branches/compare/v0.2.0..v0.1.1
[0.1.1]: https://bitbucket.org/shelacek/ubjson/branches/compare/v0.1.1..v0.1.0
[0.1.0]: https://bitbucket.org/shelacek/ubjson/commits/tag/v0.1.0


test/e2e.js
const test = require('tape');
const ubjson = require('../dist/ubjson');

test('encode/decode complex object', t => {
	const expected = {
		hello: 'world',
		from: ['UBJSON'],
		colors: [
			[255, 255, 255],
			[0, 0, 0],
			[64, 64, 96]
		],
		domains: {
			com: 'commercial',
			org: 'organization',
			net: 'network'
		},
		entires: [
			{
				id: 1,
				name: 'test',
				content: null,
				timestamp: 1532432408.008,
				published: false
			}, {
				id: 2,
				name: 'lorem',
				content: 'Lorem ipsum...',
				timestamp: 1532432416.346,
				published: true
			}
		]
	};
	const actual = ubjson.decode(ubjson.encode(
		expected,
		{ optimizeArrays: true, optimizeObjects: true }
	));
	t.deepEqual(actual, expected);
	t.end();
});


test/decode.js
const test = require('tape');
const util = require('util');
const ubjson = require('../dist/ubjson');

function toBuffer(...args) {
	return Uint8Array.from(args, x => x === +x ? x : x.charCodeAt()).buffer;
}

test('decode unsupported type', t => {
	t.throws(() => ubjson.decode(toBuffer('!')));
	t.end();
});

test('decode undefined', t => {
	t.equal(
		ubjson.decode(undefined),
		undefined
	);
	t.end();
});

test('decode null', t => {
	t.equal(
		ubjson.decode(toBuffer('Z')),
		null
	);
	t.end();
});

test('decode no-op', t => {
	t.equal(
		ubjson.decode(toBuffer('N')),
		undefined
	);
	t.end();
});

test('decode true', t => {
	t.equal(
		ubjson.decode(toBuffer('T')),
		true
	);
	t.end();
});

test('decode false', t => {
	t.equal(
		ubjson.decode(toBuffer('F')),
		false
	);
	t.end();
});

test('decode int8', t => {
	t.equal(
		ubjson.decode(toBuffer('i', 100)),
		100
	);
	t.end();
});

test('decode uint8', t => {
	t.equal(
		ubjson.decode(toBuffer('U', 200)),
		200
	);
	t.end();
});

test('decode int16', t => {
	t.equal(
		ubjson.decode(toBuffer('I', 0x12, 0x34)),
		0x1234
	);
	t.end();
});

test('decode int32', t => {
	t.equal(
		ubjson.decode(toBuffer('l', 0x12, 0x34, 0x56, 0x78)),
		0x12345678
	);
	t.end();
});

test('decode int64 [error]', t => {
	t.throws(() => ubjson.decode(
		toBuffer('L', 0x12, 0x34, 0x56, 0x78, 0x9a, 0xbc, 0xde, 0xf0)
	));
	t.end();
});

test('decode int64 [skip]', t => {
	t.doesNotThrow(() => ubjson.decode(
		toBuffer('L', 0x12, 0x34, 0x56, 0x78, 0x9a, 0xbc, 0xde, 0xf0),
		{ int64Handling: 'skip' }
	));
	t.end();
});

test('decode int64 [raw]', t => {
	t.deepEqual(
		ubjson.decode(
			toBuffer('L', 0x12, 0x34, 0x56, 0x78, 0x9a, 0xbc, 0xde, 0xf0),
			{ int64Handling: 'raw' }
		),
		[0x12, 0x34, 0x56, 0x78, 0x9a, 0xbc, 0xde, 0xf0]
	);
	t.end();
});

test('decode int64 [custom handler]', t => {
	t.deepEqual(
		ubjson.decode(
			toBuffer('L', 0x12, 0x34, 0x56, 0x78, 0x9a, 0xbc, 0xde, 0xf0),
			{
				int64Handling: ({ view }, offset) => ({
					int64: [view.getUint32(offset), view.getUint32(offset + 4)]
				})
			}
		),
		{ int64: [0x12345678, 0x9abcdef0] }
	);
	t.end();
});

test('decode int64 [invalid option]', t => {
	t.throws(() => ubjson.decode(
		toBuffer('L', 0x12, 0x34, 0x56, 0x78, 0x9a, 0xbc, 0xde, 0xf0),
		{ int64Handling: 'invalid' }
	));
	t.end();
});

test('decode float32', t => {
	t.equal(
		ubjson.decode(toBuffer('d', 0x3f, 0x80, 0x80, 0x00)),
		1.00390625
	);
	t.end();
});

test('decode float64', t => {
	t.equal(
		ubjson.decode(toBuffer('D', 0x40, 0xf8, 0x6a, 0x00, 0x10, 0x00, 0x00, 0x00)),
		100000.00390625
	);
	t.end();
});

test('decode high-precision number [error]', t => {
	t.throws(() => ubjson.decode(
		toBuffer('H', 'i', 3, '1', '.', '1')
	));
	t.end();
});

test('decode high-precision number [skip]', t => {
	t.doesNotThrow(() => ubjson.decode(
		toBuffer('H', 'i', 3, '1', '.', '1'),
		{ highPrecisionNumberHandling: 'skip' }
	));
	t.end();
});

test('decode high-precision number [raw]', t => {
	t.equal(
		ubjson.decode(
			toBuffer('H', 'i', 3, '1', '.', '1'),
			{ highPrecisionNumberHandling: 'raw' }
		),
		'1.1'
	);
	t.end();
});

test('decode high-precision number [custom handler]', t => {
	t.deepEqual(
		ubjson.decode(
			toBuffer('H', 'i', 3, '1', '.', '1'),
			{
				highPrecisionNumberHandling: ({ array }, offset, byteLength) => {
					const view = new DataView(array.buffer, offset, byteLength);
					const decoder = new util.TextDecoder();
					return { highPrecisionNumber: decoder.decode(view) };
				}
			}
		),
		{ highPrecisionNumber: '1.1' }
	);
	t.end();
});

test('decode char', t => {
	t.equal(
		ubjson.decode(toBuffer('C', 'a')),
		'a'
	);
	t.end();
});

test('decode string', t => {
	t.equal(
		ubjson.decode(toBuffer('S', 'i', 6, 'u', 'b', 'j', 's', 'o', 'n')),
		'ubjson'
	);
	t.end();
});

test('decode array', t => {
	t.deepEqual(
		ubjson.decode(toBuffer('[', 'i', 1, 'i', 2, 'i', 3, ']')),
		[1, 2, 3]
	);
	t.end();
});

test('decode array (with no-op)', t => {
	t.deepEqual(
		ubjson.decode(toBuffer('[', 'i', 1, 'N', 'i', 2, 'i', 3, 'N', ']')),
		[1, 2, 3]
	);
	t.end();
});

test('decode array (empty, optimized)', t => {
	t.deepEqual(
		ubjson.decode(toBuffer('[', '#', 'i', 0)),
		[]
	);
	t.end();
});

test('decode array (mixed, optimized)', t => {
	t.deepEqual(
		ubjson.decode(toBuffer('[', '#', 'i', 3, 'i', 1, 'C', 'a', 'T')),
		[1, 'a', true]
	);
	t.end();
});

test('decode array (strongly typed, optimized)', t => {
	t.deepEqual(
		ubjson.decode(
			toBuffer('[', '$', 'i', '#', 'i', 3, 1, 2, 3)
		),
		[1, 2, 3]
	);
	t.end();
});

test('decode array (strongly typed, unexpected eof, optimized)', t => {
	t.throws(() => ubjson.decode(toBuffer('[', '$', 'i', '#', 'i', 3, 1, 2)));
	t.end();
});

test('decode array (strongly typed, invalid length value, optimized)', t => {
	t.throws(() => ubjson.decode(toBuffer('[', '$', 'i', '#', 'i', -1)));
	t.end();
});

test('decode array (strongly typed, invalid length type, optimized)', t => {
	t.throws(() => ubjson.decode(toBuffer('[', '$', 'i', '#', 'C', '0')));
	t.end();
});

test('decode array (strongly typed, malformed, optimized)', t => {
	t.throws(() => ubjson.decode(toBuffer('[', '$', 'i', 1, 2, 3, ']')));
	t.end();
});

test('decode array (only null values, optimized)', t => {
	t.deepEqual(
		ubjson.decode(toBuffer('[', '$', 'Z', '#', 'i', 3)),
		[null, null, null]
	);
	t.end();
});

test('decode array (int8, strongly typed, optimized) [use typed array]', t => {
	const actual = ubjson.decode(
		toBuffer('[', '$', 'i', '#', 'i', 2, 0x12, 0xfe),
		{ useTypedArrays: true }
	);
	t.assert(actual.constructor.name === 'Int8Array');
	t.deepEqual(actual, [18, -2]);
	t.end();
});

test('decode array (uint8, strongly typed, optimized) [use typed array]', t => {
	const actual = ubjson.decode(
		toBuffer('[', '$', 'U', '#', 'i', 2, 0x12, 0xfe),
		{ useTypedArrays: true }
	);
	t.assert(actual.constructor.name === 'Uint8Array');
	t.deepEqual(actual, [18, 254]);
	t.end();
});

test('decode array (int16, strongly typed, optimized) [use typed array]', t => {
	const actual = ubjson.decode(
		toBuffer('[', '$', 'I', '#', 'i', 2, 0x12, 0x34, 0xfe, 0xdc),
		{ useTypedArrays: true }
	);
	t.assert(actual.constructor.name === 'Int16Array');
	t.deepEqual(actual, [4660, -292]);
	t.end();
});

test('decode array (int32, strongly typed, optimized) [use typed array]', t => {
	const actual = ubjson.decode(
		toBuffer(
			'[', '$', 'l', '#', 'i', 2,
			0x12, 0x34, 0x56, 0x78, 0xfe, 0xdc, 0xba, 0x98
		),
		{ useTypedArrays: true }
	);
	t.assert(actual.constructor.name === 'Int32Array');
	t.deepEqual(actual, [305419896, -19088744]);
	t.end();
});

test('decode array (float32, strongly typed, optimized) [use typed array]', t => {
	const actual = ubjson.decode(
		toBuffer(
			'[', '$', 'd', '#', 'i', 2,
			0x3e, 0x80, 0x00, 0x00, 0x3e, 0x00, 0x00, 0x00
		),
		{ useTypedArrays: true }
	);
	t.assert(actual.constructor.name === 'Float32Array');
	t.deepEqual(actual, [0.25, 0.125]);
	t.end();
});

test('decode array (float64, strongly typed, optimized) [use typed array]', t => {
	const actual = ubjson.decode(
		toBuffer(
			'[', '$', 'D', '#', 'i', 2,
			0x3f, 0xd0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
			0x3f, 0xc0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
		),
		{ useTypedArrays: true }
	);
	t.assert(actual.constructor.name === 'Float64Array');
	t.deepEqual(actual, [0.25, 0.125]);
	t.end();
});

test('decode object', t => {
	t.deepEqual(
		ubjson.decode(toBuffer(
			'{',
			'i', 1, 'a', 'i', 1,
			'i', 1, 'b', 'i', 2,
			'i', 1, 'c', 'i', 3,
			'}'
		)),
		{ a: 1, b: 2, c: 3 }
	);
	t.end();
});

test('decode object (with no-op)', t => {
	t.deepEqual(
		ubjson.decode(toBuffer(
			'N', '{',
			'N', 'i', 1, 'a', 'i', 1,
			'i', 1, 'b', 'N', 'i', 2,
			'i', 1, 'c', 'i', 3, 'N',
			'}', 'N'
		)),
		{ a: 1, b: 2, c: 3 }
	);
	t.end();
});

test('decode array (empty, optimized)', t => {
	t.deepEqual(
		ubjson.decode(toBuffer('{', '#', 'i', 0)),
		{}
	);
	t.end();
});

test('decode object (mixed, optimized)', t => {
	t.deepEqual(
		ubjson.decode(toBuffer(
			'{', '#', 'i', 3,
			'i', 1, 'a', 'i', 1,
			'i', 1, 'b', 'C', 'a',
			'i', 1, 'c', 'T'
		)),
		{ a: 1, b: 'a', c: true }
	);
	t.end();
});

test('decode object (strongly typed, optimized)', t => {
	t.deepEqual(
		ubjson.decode(toBuffer(
			'{', '$', 'i', '#', 'i', 3,
			'i', 1, 'a', 1,
			'i', 1, 'b', 2,
			'i', 1, 'c', 3
		)),
		{ a: 1, b: 2, c: 3 }
	);
	t.end();
});

test('decode object (only null values, optimized)', t => {
	t.deepEqual(
		ubjson.decode(toBuffer(
			'{', '$', 'Z', '#', 'i', 3,
			'i', 1, 'a',
			'i', 1, 'b',
			'i', 1, 'c'
		)),
		{ a: null, b: null, c: null }
	);
	t.end();
});


test/encode.js
const test = require('tape');
const ubjson = require('../dist/ubjson');

function toArray(...args) {
	if (args[0] instanceof ArrayBuffer) {
		return Array.from(new Uint8Array(args[0]));
	}
	return args.map(x => x === +x ? x : x.charCodeAt());
}

test('encode unsupported value', t => {
	// eslint-disable-next-line prefer-arrow-callback
	t.throws(() => ubjson.encode(function () {}));
	t.end();
});

test('encode undefined', t => {
	t.deepEqual(
		toArray(ubjson.encode(undefined)),
		toArray('N')
	);
	t.end();
});

test('encode null', t => {
	t.deepEqual(
		toArray(ubjson.encode(null)),
		toArray('Z')
	);
	t.end();
});

test('encode true', t => {
	t.deepEqual(
		toArray(ubjson.encode(true)),
		toArray('T')
	);
	t.end();
});

test('encode false', t => {
	t.deepEqual(
		toArray(ubjson.encode(false)),
		toArray('F')
	);
	t.end();
});

test('encode int8', t => {
	t.deepEqual(
		toArray(ubjson.encode(100)),
		toArray('i', 100)
	);
	t.end();
});

test('encode uint8', t => {
	t.deepEqual(
		toArray(ubjson.encode(200)),
		toArray('U', 200)
	);
	t.end();
});

test('encode int16', t => {
	t.deepEqual(
		toArray(ubjson.encode(0x1234)),
		toArray('I', 0x12, 0x34)
	);
	t.end();
});

test('encode int32', t => {
	t.deepEqual(
		toArray(ubjson.encode(0x12345678)),
		toArray('l', 0x12, 0x34, 0x56, 0x78)
	);
	t.end();
});

test('encode float32', t => {
	t.deepEqual(
		toArray(ubjson.encode(1.00390625)),
		toArray('d', 0x3f, 0x80, 0x80, 0x00)
	);
	t.end();
});

test('encode float32 (too large integer)', t => {
	t.deepEqual(
		toArray(ubjson.encode(2147483648)),
		toArray('d', 0x4f, 0x00, 0x00, 0x00)
	);
	t.end();
});

test('encode float64', t => {
	t.deepEqual(
		toArray(ubjson.encode(100000.00390625)),
		toArray('D', 0x40, 0xf8, 0x6a, 0x00, 0x10, 0x00, 0x00, 0x00)
	);
	t.end();
});

test('encode char', t => {
	t.deepEqual(
		toArray(ubjson.encode('a')),
		toArray('C', 'a')
	);
	t.end();
});

test('encode string', t => {
	t.deepEqual(
		toArray(ubjson.encode('ubjson')),
		toArray('S', 'i', 6, 'u', 'b', 'j', 's', 'o', 'n')
	);
	t.end();
});

test('encode array', t => {
	t.deepEqual(
		toArray(ubjson.encode([1, 2, 3])),
		toArray('[', 'i', 1, 'i', 2, 'i', 3, ']')
	);
	t.end();
});

test('encode array (int8) [only typed array]', t => {
	t.deepEqual(
		toArray(ubjson.encode([1, 2, 3], { optimizeArrays: 'onlyTypedArrays' })),
		toArray('[', 'i', 1, 'i', 2, 'i', 3, ']')
	);
	t.end();
});

test('encode array (empty)  [optimize]', t => {
	t.deepEqual(
		toArray(ubjson.encode([], { optimizeArrays: true })),
		toArray('[', '#', 'i', 0)
	);
	t.end();
});

test('encode array (mixed) [optimize]', t => {
	t.deepEqual(
		toArray(ubjson.encode([1, 'a', true], { optimizeArrays: true })),
		toArray('[', '#', 'i', 3, 'i', 1, 'C', 'a', 'T')
	);
	t.end();
});

test('encode array (int8) [optimize]', t => {
	t.deepEqual(
		toArray(ubjson.encode([1, 2, 3], { optimizeArrays: true })),
		toArray('[', '$', 'i', '#', 'i', 3, 1, 2, 3)
	);
	t.end();
});

test('encode array (int16) [optimize]', t => {
	t.deepEqual(
		toArray(ubjson.encode([255, -1], { optimizeArrays: true })),
		toArray('[', '$', 'I', '#', 'i', 2, 0x00, 0xff, 0xff, 0xff)
	);
	t.end();
});

test('encode array (only null values) [optimize]', t => {
	t.deepEqual(
		toArray(ubjson.encode([null, null, null], { optimizeArrays: true })),
		toArray('[', '$', 'Z', '#', 'i', 3)
	);
	t.end();
});

test('encode array (uint8 typed array)', t => {
	t.deepEqual(
		toArray(ubjson.encode(Uint8Array.from([1, 2]))),
		toArray('[', 'i', 1, 'i', 2, ']')
	);
	t.end();
});

test('encode array (int8 typed array) [only typed array]', t => {
	t.deepEqual(
		toArray(ubjson.encode(
			Int8Array.from([18, -2]),
			{ optimizeArrays: 'onlyTypedArrays' }
		)),
		toArray('[', '$', 'i', '#', 'i', 2, 0x12, 0xfe)
	);
	t.end();
});

test('encode array (uint8 typed array) [only typed array]', t => {
	t.deepEqual(
		toArray(ubjson.encode(
			Uint8Array.from([18, 254]),
			{ optimizeArrays: 'onlyTypedArrays' }
		)),
		toArray('[', '$', 'U', '#', 'i', 2, 0x12, 0xfe)
	);
	t.end();
});

test('encode array (int16 typed array) [only typed array]', t => {
	t.deepEqual(
		toArray(ubjson.encode(
			Int16Array.from([4660, -292]),
			{ optimizeArrays: 'onlyTypedArrays' }
		)),
		toArray('[', '$', 'I', '#', 'i', 2, 0x12, 0x34, 0xfe, 0xdc)
	);
	t.end();
});

test('encode array (int32 typed array) [only typed array]', t => {
	t.deepEqual(
		toArray(ubjson.encode(
			Int32Array.from([305419896, -19088744]),
			{ optimizeArrays: 'onlyTypedArrays' }
		)),
		toArray('[', '$', 'l', '#', 'i', 2, 0x12, 0x34, 0x56, 0x78, 0xfe, 0xdc, 0xba, 0x98)
	);
	t.end();
});

test('encode array (float32 typed array) [only typed array]', t => {
	t.deepEqual(
		toArray(ubjson.encode(
			Float32Array.from([0.25, 0.125]),
			{ optimizeArrays: 'onlyTypedArrays' }
		)),
		toArray('[', '$', 'd', '#', 'i', 2, 0x3e, 0x80, 0x00, 0x00, 0x3e, 0x00, 0x00, 0x00)
	);
	t.end();
});

test('encode array (float64 typed array) [only typed array]', t => {
	t.deepEqual(
		toArray(ubjson.encode(
			Float64Array.from([0.25, 0.125]),
			{ optimizeArrays: 'onlyTypedArrays' }
		)),
		toArray(
			'[', '$', 'D', '#', 'i', 2,
			0x3f, 0xd0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
			0x3f, 0xc0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
		)
	);
	t.end();
});

test('encode array (uint8 typed array) [optimize]', t => {
	t.deepEqual(
		toArray(ubjson.encode(Uint8Array.from([1, 2, 3]), { optimizeArrays: true })),
		toArray('[', '$', 'U', '#', 'i', 3, 1, 2, 3)
	);
	t.end();
});

test('encode object', t => {
	t.deepEqual(
		toArray(ubjson.encode({ a: 1, b: 2, c: 3 })),
		toArray(
			'{',
			'i', 1, 'a', 'i', 1,
			'i', 1, 'b', 'i', 2,
			'i', 1, 'c', 'i', 3,
			'}'
		)
	);
	t.end();
});

test('encode object (empty)  [optimize]', t => {
	t.deepEqual(
		toArray(ubjson.encode({}, { optimizeObjects: true })),
		toArray('{', '#', 'i', 0)
	);
	t.end();
});

test('encode object (mixed) [optimize]', t => {
	t.deepEqual(
		toArray(ubjson.encode({ a: 1, b: 'a', c: true }, { optimizeObjects: true })),
		toArray(
			'{', '#', 'i', 3,
			'i', 1, 'a', 'i', 1,
			'i', 1, 'b', 'C', 'a',
			'i', 1, 'c', 'T'
		)
	);
	t.end();
});

test('encode object (strongly typed) [optimize]', t => {
	t.deepEqual(
		toArray(ubjson.encode({ a: 1, b: 2, c: 3 }, { optimizeObjects: true })),
		toArray(
			'{', '$', 'i', '#', 'i', 3,
			'i', 1, 'a', 1,
			'i', 1, 'b', 2,
			'i', 1, 'c', 3
		)
	);
	t.end();
});

test('encode object (only null values) [optimize]', t => {
	t.deepEqual(
		toArray(ubjson.encode({ a: null, b: null, c: null }, { optimizeObjects: true })),
		toArray(
			'{', '$', 'Z', '#', 'i', 3,
			'i', 1, 'a',
			'i', 1, 'b',
			'i', 1, 'c'
		)
	);
	t.end();
});


src/ubjson-encoder.js
export class UbjsonEncoder {
	constructor(options = {}) {
		this._options = options;
		this._textEncoder = new (typeof TextEncoder !== 'undefined'
			? TextEncoder
			: require('util').TextEncoder // eslint-disable-line global-require
		)();
	}

	encode(value) {
		const packers = this._encode(value);
		const size = packers.reduce((acc, x) => acc + x.byteLength, 0);
		const array = new Uint8Array(size);
		const view = new DataView(array.buffer);
		const storage = { array, view };
		let offset = 0;
		for (const packer of packers) {
			packer.storer(storage, offset);
			offset += packer.byteLength;
		}
		return array.buffer;
	}

	_encode(value) {
		const type = this._getType(value);
		return [
			this._packType(type),
			...this._encodeValue(value, type)
		];
	}

	_encodeValue(value, type) {
		let char;
		let bytes;
		switch (type) {
			case 'C':
				char = value.charCodeAt();
				return [this._pack(({ view }, offset) => view.setInt8(offset, char), 1)];
			case 'S':
				bytes = this._textEncoder.encode(value);
				return [
					...this._encode(bytes.byteLength),
					this._packTypedArray(bytes)
				];
			case 'i':
				return [this._pack(({ view }, offset) => view.setInt8(offset, value), 1)];
			case 'U':
				return [this._pack(({ view }, offset) => view.setUint8(offset, value), 1)];
			case 'I':
				return [this._pack(({ view }, offset) => view.setInt16(offset, value), 2)];
			case 'l':
				return [this._pack(({ view }, offset) => view.setInt32(offset, value), 4)];
			case 'd':
				return [this._pack(({ view }, offset) => view.setFloat32(offset, value), 4)];
			case 'D':
				return [this._pack(({ view }, offset) => view.setFloat64(offset, value), 8)];
			case '[':
				return this._encodeArray(value);
			case '{':
				return this._encodeObject(value);
		}
		return [];
	}

	_getType(value) {
		if (value === null) {
			return 'Z';
		}
		switch (typeof value) {
			case 'undefined':
				return 'N';
			case 'boolean':
				return value ? 'T' : 'F';
			case 'string':
				return (value.length === 1 && value.charCodeAt() <= 127) ? 'C' : 'S';
			case 'number':
				if (Number.isInteger(value)) {
					if (-128 <= value && value <= 127) {
						return 'i';
					}
					if (0 <= value && value <= 255) {
						return 'U';
					}
					if (-32768 <= value && value <= 32767) {
						return 'I';
					}
					if (-2147483648 <= value && value <= 2147483647) {
						return 'l';
					}
				}
				return (Number.isNaN(value) || Math.fround(value) === value) ? 'd' : 'D';
			case 'object':
				return Array.isArray(value) || ArrayBuffer.isView(value) ? '[' : '{';
		}
		throw new Error('Value cannot be serialized');
	}

	_encodeArray(value) {
		let type;
		if ((this._options.optimizeArrays === true
				|| this._options.optimizeArrays === 'onlyTypedArrays')
				&& ArrayBuffer.isView(value)) {
			switch (value.constructor.name) {
				case 'Int8Array':
					return [].concat(
						this._packContainerMarkers('i', value.length),
						this._packTypedArray(value)
					);
				case 'Uint8Array':
					return [].concat(
						this._packContainerMarkers('U', value.length),
						this._packTypedArray(value)
					);
				case 'Int16Array':
					type = 'I';
					break;
				case 'Int32Array':
					type = 'l';
					break;
				case 'Float32Array':
					type = 'd';
					break;
				case 'Float64Array':
					type = 'D';
					break;
			}
		}
		const items = (Array.isArray(value) ? value : Array.from(value)).map(x => ({
			type: type || this._getType(x),
			value: x
		}));
		return this._packContainerItems(items, ']', type || this._options.optimizeArrays === true);
	}

	_encodeObject(value) {
		const items = Object.entries(value).map(x => ({
			key: x[0],
			type: this._getType(x[1]),
			value: x[1]
		}));
		return this._packContainerItems(items, '}', this._options.optimizeObjects === true);
	}

	_packContainerMarkers(type, count) {
		const packers = [];
		if (count != null) {
			if (type) {
				packers.push(this._packType('$'), this._packType(type));
			}
			packers.push(this._packType('#'), ...this._encode(count));
		}
		return packers;
	}

	_packContainerItems(items, terminator, optimize) {
		let type;
		let count;
		if (optimize) {
			if (items.length) {
				type = this._obtainCommonType(items);
			}
			count = items.length;
		}
		const packers = this._packContainerMarkers(type, count);
		for (const item of items) {
			item.key && packers.push(...this._encodeValue(item.key, 'S'));
			!type && packers.push(this._packType(item.type));
			packers.push(...this._encodeValue(item.value, type || item.type));
		}
		if (count == null) {
			packers.push(this._packType(terminator));
		}
		return packers;
	}

	_obtainCommonType(items) {
		const type = items.map(x => x.type).reduce(this._typeReducer);
		return (type === 'U' && items.some(x => x.value < 0)) ? 'I' : type;
	}

	_typeReducer(a, b) {
		if (a === b) {
			return a;
		}
		if (!a) {
			return null;
		}
		const reduceTo = seq => seq[Math.min(seq.indexOf(a), seq.indexOf(b))];
		return reduceTo('Dd') || reduceTo('SC') || reduceTo('lIUi');
	}

	_packType(value) {
		return this._pack(({ view }, offset) => view.setInt8(offset, value.charCodeAt()), 1);
	}

	_packTypedArray(value) {
		return this._pack(({ array }, offset) => array.set(value, offset), value.byteLength);
	}

	_pack(storer, byteLength) {
		return { storer, byteLength };
	}
}
